MINOR: Fixed a division by 0 scenario

TimYagan opened the pull request:
MINOR: Fixed a division by 0 scenario by ensuring that totalCapacity had a default value of 1 in the event that sumCapacity was 0, thereby negating a division by 0 event which could have caused a denial of service.

This was tested via static analysis while I was reviewing the code. There is no need for additional unit tests since this is indeed a minor change.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

TimYagan committed:
MINOR: Fixed a division by 0 scenario

MINOR: Fixed a division by 0 scenario by ensuring that totalCapacity had a default value of 1 in the event that sumCapacity was 0, thereby negating a division by 0 event which could have caused a denial of service.

This was tested via static analysis while I was reviewing the code. There is no need for additional unit tests since this is indeed a minor change.

ryannedolan commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:
```diff
@@ -72,7 +72,13 @@ private void assignStandby(final int numStandbyReplicas) {
+        final int totalCapacity = 1;
+        
+        if(sumCapacity(clients.value()) > 0) {
+            totalCapacity = sumCapacity(clients.value();
```
reassignment to a final? Recommend:

```
final int totalCapacity;
if (...) {
   totalCapacity = ...
} else {
  totalCapacity = ...
}
```

TimYagan replied:
Nice catch!

bbejeck replied:
@TimYagan thanks for the patch!  While I see the point here, if there's a situation the capacity across all clients is `0`when initially assigning tasks, it's probably indicative of a broader issue and maybe we should throw an `Exception` vs. setting a value of `1`.  

\cc @mjsax WDYT?

TimYagan committed:
Changed the code to avoid reassigning a final

Changed the code to avoid reassigning a final, after the review. Thanks for catching that.

TimYagan commented:
Yeah, throwing an exception may be more appropriate here, especially if
there is an issue in other classes. Speaking of which, perhaps we should
have a test that could also catch that scenario if ever it were to happen?

On Tue, 1 Oct 2019, 16:45 Bill Bejeck, <notifications@github.com> wrote:

> *@bbejeck* commented on this pull request.
> ------------------------------
>
> In
> streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
> <https://github.com/apache/kafka/pull/7414#discussion_r330098219>:
>
> > @@ -72,7 +72,13 @@ private void assignStandby(final int numStandbyReplicas) {
>      }
>
>      private void assignActive() {
> -        final int totalCapacity = sumCapacity(clients.values());
> +
> +        final int totalCapacity = 1;
> +
> +        if(sumCapacity(clients.value()) > 0) {
> +            totalCapacity = sumCapacity(clients.value();
>
> @TimYagan <https://github.com/TimYagan> thanks for the patch! While I see
> the point here, if there's a situation the capacity across all clients is
> 0when initially assigning tasks, it's probably indicative of a broader
> issue and maybe we should throw an Exception vs. setting a value of 1.
>
> \cc @mjsax <https://github.com/mjsax> WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/7414?email_source=notifications&email_token=AHMKZKFHU3ITUJNKVLGIYGTQMNPATA5CNFSM4I3TSPZ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGPWOVY#discussion_r330098219>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHMKZKBQXVQKE43SK2MWP63QMNPATANCNFSM4I3TSPZQ>
> .
>

mjsax commented:
I agree with @bbejeck -- the client capacity should never be zero and neither should the total capacity.

Hence, it might be best to actually check `client.capacity() > 0` within `sumCapacity` (and maybe if the number of clients is larger than 1)?

mjsax commented:
Closing this PR due to inactivity -- feel free to re-open. However, there is not really a divide-by-zero issue. We could add a comment to clarify this though.

big-andy-coates commented:
FYI, this is probably being fixed because of this: https://ossindex.sonatype.org/vulnerability/sonatype-2019-0422?component-type=maven&component-name=org.apache.kafka%2Fkafka-streams&utm_source=ossindex-client&utm_medium=integration&utm_content=1.7.0

mjsax commented:
Addressed via https://github.com/apache/kafka/pull/13175
